### ISSUE TITLE
Dismiss popups via click away

### ISSUE THREAD

**Issue by zoriya** [Aylur/ags#63]
Clicking out of a popup window should close it. The eww way of doing this is to create an invisible button the size of the screen and close the popup when this button is pressed but I would prefer having it automatically by ags.

**Comment by Aylur** [comment]
This is still how I do it and how it should be done imo, window anchor doesnt play nice with this approach, there are 2^4 combinations of how a window could be positioned, and you can't just have one EventBox behind it to do this, because clicking inside the window would propagate to the eventbox.
But if you find a simple solution to this it could be a great addition

**Comment by DevAlien** [comment]
What about creating 2 windows? 1 full and then one the size you want?

**Comment by Yaikava** [comment]
Does anyone have some simple example for this? I am not that good with css and not really sure how I would do this.

**Comment by eyezahhhh** [comment]
This is what I use
```js
const WindowCloser = (menu, monitor) => {
    const name = `WindowCloser-${menu}`;

    function close() {
        ags.App.closeWindow(menu);
        ags.App.closeWindow(name);
    }

    return Window({
        name,
        className: 'windowCloser',
        layer: 'top',
        monitor,
        popup: true,
        anchor: ['top', 'bottom', 'left', 'right'],
        child: EventBox({
            onPrimaryClick: close
        })
    })
}


export default {
    windows: [
        WindowCloser('StartMenu', 0),
        StartMenu
    ],
};
```

And to toggle the menu:
```js
ags.App.toggleWindow('StartMenu');
ags.App.toggleWindow('WindowCloser-StartMenu');
```

No idea if it's a good implementation but it works for me. I create a new instance of `WindowCloser` for each window that I want to be able to close by clicking away.

**Comment by webflo-dev** [comment]
There is a simplier way to do it:
```javascript
export const PopupWindow = (
  { name, popup = true, focusable = true, anchor = [] },
  content,
) => {
  const _window = Widget.Window({
    name,
    popup,
    focusable,
    anchor,
    visible: false,
    child: content,
  });

  _window.connect("button-release-event", (emitter) => {
    const [x, y] = emitter.get_pointer();
    if (x === 0 && y === 0) {
      App.closeWindow(name);
    }
  });

  return _window;
};
```

**Comment by Aylur** [comment]
I think either DevAlien or kotontrion posted this on the discord server, and for some reason it did not occur to me that the signal is still emitted when you click outside the window, and so I didn't even test this, but now I realized that this is the solution for this.

**Comment by Aylur** [comment]
you also need `focusable: true` in order for the signal though, probably thats why it didn't occur to me

**Comment by its-darsh** [comment]
TL;DR: bad idea

if gtk doesn't help doing that you can use libinput to catch pointer clicks and if there is a click that have been done in libinput but haven't caught by gtk so this is a click happened outside of your widgets

**Comment by ldelossa** [comment]
For some reason, your solution @Aylur does not seem to work on Sway. Do you happen to have an idea why?

**Comment by Aylur** [comment]
this hack was patched in Hyprland too
```js
  _window.connect("button-release-event", (emitter) => {
    const [x, y] = emitter.get_pointer();
    if (x === 0 && y === 0) {
      App.closeWindow(name);
    }
  });
```

I think we can say goodbye to `popup` on Window widgets
#248 will take care of the close on ESC feature
and for clickaway the end user will have to implement a solution, so that we can stay compositor agnostic

**Comment by cafetestrest** [comment]
@Aylur does this mean that as of recent Hyprland update clicking outside of the widget won't close it?

**Comment by daUnknownCoder** [comment]
> @Aylur does this mean that as of recent Hyprland update clicking outside of the widget won't close it?

yeah i lose input focus if i (click outside of widget which doesnt close it)/(click on widget again to close it)

**Comment by daUnknownCoder** [comment]
Any workaround @Aylur?

**Comment by musjj** [comment]
Wouldn't it be possible to detect the focus of the window itself. So when it gets unfocused, it will close itself.

I can't figure out how to make it work though, `notify::is-focus` doesn't seem to get triggered.

**Comment by Aylur** [comment]
two workaround I know works
1. a "keygrabber" window opened behind the popup window which is a single fullscreen eventbox that closes itself and the popup on click
2. making the popup window fullscreen and padding it out with eventboxes that close on click
